### PR TITLE
=str #16997 added variance annotation to UniformFanInShape

### LINKED ISSUE #16997: No covariance in graph ~> operator
This does not compile
``` scala
FlowGraph.closed() { implicit b =>
  val merge = b.add(Merge[AnyRef](inputPorts = 2))
  Source.single("") ~> merge // compile error
  Source.single('a) ~> merge // compile error
  merge ~> Sink.head[AnyRef]()
}
```

while this does
``` scala
FlowGraph.closed() { implicit b =>
  val merge = b.add(Merge[AnyRef](inputPorts = 2))
  Source.single("") ~> merge.in(0)
  Source.single('a) ~> merge.in(1)
  merge ~> Sink.head[AnyRef]()
}
```

It seems that covariance is missing on the ~> type parameters in  https://github.com/akka/akka/blob/ab5702e65ce1e78a88bc448b4ca2798464ce84d3/akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala#L289

### PR DESCRIPTION
**Pull request by 2m**
Partially fixes #16997.

I was unable to add variance annotation to `UniformFanOutShape`, beucase `I` type argument there is used in invariant positions when extending and in `construct` method.

### CODE REVIEW

File: akka-stream/src/main/boilerplate/akka/stream/FanInShape.scala.template [hunk 1]
```diff
@@ -46,15 +46,15 @@ object UniformFanInShape {
...
   def this(n: Int) = this(n, FanInShape.Name("UniformFanIn"))
   def this(n: Int, name: String) = this(n, FanInShape.Name(name))
   def this(outlet: Outlet[O], inlets: Array[Inlet[T]]) = this(inlets.length, FanInShape.Ports(outlet, inlets.toList))
   override protected def construct(init: FanInShape.Init[O]): FanInShape[O] = new UniformFanInShape(n, init)
   override def deepCopy(): UniformFanInShape[T, O] = super.deepCopy().asInstanceOf[UniformFanInShape[T, O]]
-  
-  val inArray: Array[Inlet[T]] = Array.tabulate(n)(i => newInlet[T](s"in$i"))
-  def in(n: Int): Inlet[T] = inArray(n)
+
+  val inSeq: immutable.IndexedSeq[Inlet[T]] = Vector.tabulate(n)(i => newInlet[T](s"in$i"))
```
Comment (drewhk) [inline akka-stream/src/main/boilerplate/akka/stream/FanInShape.scala.template]: This makes me sad :( Unfortunately arrays are invariant, so I guess this is how it is.

**Comment by drewhk** [comment]
LGTM

File: akka-stream/src/main/boilerplate/akka/stream/FanInShape.scala.template (hunk 1 above)
Comment (rkuhn) [inline akka-stream/src/main/boilerplate/akka/stream/FanInShape.scala.template]: In a perfect universe we’d have a properly covariant ImmutableArray.

Comment (ktoso) [inline akka-stream/src/main/boilerplate/akka/stream/FanInShape.scala.template]: So... when do we write a new VM?

Comment (drewhk) [inline akka-stream/src/main/boilerplate/akka/stream/FanInShape.scala.template]: I'm in

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowGraphCompileSpec.scala [hunk 2]
```diff
@@ -224,6 +224,38 @@ class FlowGraphCompileSpec extends AkkaSpec {
...
+        Source[Apple](apples) ~> appleMerge
+        appleMerge ~> Sink.head[Fruit]
+
+        val appleMerge2 = b.add(Merge[Apple](1))
+        Source[Apple](apples) ~> appleMerge2
+        appleMerge2 ~> Sink.head[Apple]
+
+        val fruitBcast = b.add(Broadcast[Fruit](1))
+        Source[Fruit](apples) ~> fruitBcast
+        //Source[Apple](apples) ~> fruitBcast // FIXME: should compile #16997
```
Comment (ktoso) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowGraphCompileSpec.scala]: what about that FIXME? Issue nr is the one this PR solves, right?

File: akka-stream/src/main/boilerplate/akka/stream/FanInShape.scala.template (hunk 1 above)
Comment (2m) [inline akka-stream/src/main/boilerplate/akka/stream/FanInShape.scala.template]: A new VM for a

> properly covariant ImmutableArray

? That escalated quickly.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowGraphCompileSpec.scala (hunk 2 above)
Comment (ktoso) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowGraphCompileSpec.scala]: Ah, ok now I noticed the first comment on this PR.

Comment (2m) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowGraphCompileSpec.scala]: Yup. I better add a comment in the original issue.

**Comment by ktoso** [comment]
LGTM

**Comment by 2m** [comment]
Failure was TlsSpec.

**Pull request merged**